src/Interpreters/ThreadStatusExt.cpp

```diff
@@ -61,10 +61,25 @@ std::vector<UInt64> ThreadGroup::getInvolvedThreadIds() const
     return res;
 }
 
-void ThreadGroup::linkThread(UInt64 thread_it)
+size_t ThreadGroup::getPeakThreadsUsage() const
 {
     std::lock_guard lock(mutex);
-    thread_ids.insert(thread_it);
+    return peak_threads_usage;
+}
+
+
+void ThreadGroup::linkThread(UInt64 thread_id)
+{
+    std::lock_guard lock(mutex);
+    thread_ids.insert(thread_id);
+    if (active_thread_ids.insert(thread_id).second)
+        peak_threads_usage = std::max(peak_threads_usage, active_thread_ids.size());
```

Comment:
`active_thread_ids` may not be required if `thread_id` is always expected to be unique.

And it would be enough to:
`++peak_threads_usage`

And unlinkThread in `--++peak_threads_usage`.